kge/model/embedder/lookup_embedder.py

```diff
@@ -86,7 +87,8 @@ def embed(self, indexes: Tensor) -> Tensor:
         return self._embed(self.embeddings(indexes.long()))
 
     def embed_all(self) -> Tensor:
-        return self._embed(self.embeddings.weight)
+        self.all_indexes = self.all_indexes.to(self.embeddings.weight.device)
+        return self.embed(self.all_indexes)
```

Comment:
I think this would be handled much better (computation, storage) if a `None` argument was interpreted as "all_indexes" my `embed`.

Reply:
I don't understand how this would improve computation and storage, but I can shift it over to embed. 

so basically you propose:

```
    def embed(self, indexes: Tensor) -> Tensor:
        if indexes is None:
            self.all_indexes = self.all_indexes.to(self.embeddings.weight.device)
            indexes = self.all_indexes 
        return self._embed(self.embeddings(indexes.long()))

```

Reply:
No. Never use all_indexes (do not even create it). Change the API such that indexes=None means all indexes. In this case, we should return self.embeddings.weight. Or am I missing something? It's a waste of space to store a vector of 1...num_embeddings on the GPU.

Reply:
Accessing self.embeddings.weight was the problem that is being fixed here. The point of the change is that we don't get sparse gradients when we directly access self.embeddings.weight. Please note that sparse gradients is not achieved by self.embeddings.weight.to_sparse(), only the gradients are sparse. 

However, this is fixing the case of using SparseAdam+unweighted penalty, SparseAdam+1vsAll or SparseAdam+KvsAll. One could argue that those are non-sensical anyway because sparse is most likely used when we do not want to access the whole entity embedding table, i.e., for sampled training strategies and weighted penalties. We could just opt to not support that combination and not waste those 7MB / 1M entities of long indices.

Reply:
I think we should either use `.weight` or `.embeddings()` (i.e., forward), but not mix both. 

I am fine to consistently use `.embeddings()`. But then write `torch.arange(..., device=..., dtype=torch.long)` to get all indexes. I don't think there is any benefit of caching it.

Agreed?

Reply:
Ok, I thought allocating the exact same tensor constantly is more wasteful, also, with having the GPU-memory fragmentation issue in mind ... . But I'll trust your judgment.

Reply:
Let's keep it simple and only add caching etc. when we run into problems.